File: backend/app/features/twin/schemas/ai_software.py
"""Schemas for known AI software inventory from Redis twin."""

from __future__ import annotations

from typing import List, Optional

from pydantic import BaseModel, Field


class AiSoftwareItem(BaseModel):
    model_config = {"protected_namespaces": ()}

    id: str
    product_id: str = ""
    product_name: str = ""
    vendor: str = ""
    category: str = ""
    confidence: str = ""
    confidence_reason: str = ""
    installed: bool = False
    running: bool = False
    path: str = ""
    version: str = ""
    bundle_id: str = ""
    executable: str = ""
    signing_id: str = ""
    team_id: str = ""
    signature_valid: Optional[bool] = None
    matched_evidence: List[str] = Field(default_factory=list)
    failed_evidence: List[str] = Field(default_factory=list)
    # CLI agent fields (empty for GUI .app inventory).
    invocation_path: str = ""
    resolved_path: str = ""
    package_manager: str = ""
    package_identifier: str = ""
    entry_point: str = ""
    interpreter: str = ""
    # Local model runtime fields (empty/false for apps and CLI agents).
    serving: bool = False
    exposure: str = ""
    listeners: List[dict] = Field(default_factory=list)
    models_available: int = 0
    model_format: str = ""
    runtime_version: str = ""
    local_clients: List[dict] = Field(default_factory=list)


class AiSoftwareListResponse(BaseModel):
    device_id: str
    total: int = 0
    items: List[AiSoftwareItem] = Field(default_factory=list)

File: frontend/src/features/twin/types/aiSoftware.ts
export type AiSoftwareListener = {
  addr?: string;
  port?: number;
  protocol?: string;
};

export type AiSoftwareLocalClient = {
  pid?: number;
  executable?: string;
  product_id?: string;
};

export type AiSoftwareItem = {
  id: string;
  product_id: string;
  product_name: string;
  vendor: string;
  category: string;
  confidence: string;
  confidence_reason?: string;
  installed: boolean;
  running: boolean;
  path: string;
  version: string;
  bundle_id: string;
  executable: string;
  signing_id: string;
  team_id: string;
  signature_valid?: boolean | null;
  matched_evidence?: string[];
  failed_evidence?: string[];
  /** CLI agent fields (empty for GUI .app inventory). */
  invocation_path?: string;
  resolved_path?: string;
  package_manager?: string;
  package_identifier?: string;
  entry_point?: string;
  interpreter?: string;
  /** Local model runtime fields. */
  serving?: boolean;
  exposure?: string;
  listeners?: AiSoftwareListener[];
  models_available?: number;
  model_format?: string;
  runtime_version?: string;
  local_clients?: AiSoftwareLocalClient[];
  /** IDE extension fields. */
  extension_id?: string;
  host_ide_product_id?: string;
  host_ide_path?: string;
  profile?: string;
  /** null/undefined = unknown */
  enabled?: boolean | null;
  active?: boolean | null;
  mcp_configured?: boolean;
  local_model_product_id?: string;
};

export type AiSoftwareListResponse = {
  device_id: string;
  total: number;
  items: AiSoftwareItem[];
};
